File: convert-lua-string-to-yaml.py
import argparse
import re
import yaml

def parse_lua_table(lua_string):
    """Parse a Lua table string into a Python dictionary."""
    # Remove 'exportData =' and any trailing whitespace
    lua_string = re.sub(r'exportData\s*=\s*', '', lua_string).strip()
    
    def convert_value(val):
        """Convert Lua value to Python equivalent."""
        val = val.strip()
        if val.isdigit():
            return int(val)
        if val == "N/A":
            return val
        return val.strip('"')
    
    def parse_table(text, depth=0):
        """Recursively parse Lua table into Python dict."""
        result = {}
        text = text.strip('{} \n')
        if not text:
            return result
        
        # Match key-value pairs
        pattern = r'(\[?"[^"]+"\]|\d+)\s*=\s*({[^}]*}|[^,]+)(?:,|$|\s*(?=\n*\}))'
        matches = re.finditer(pattern, text, re.DOTALL)
        
        for match in matches:
            key, value = match.groups()
            # Clean up key
            if key.startswith('[') and key.endswith(']'):
                key = key[2:-2]  # Remove [""] and quotes
            else:
                key = int(key) if key.isdigit() else key
            
            # Parse value
            value = value.strip()
            if value.startswith('{'):
                # Nested table: parse key-value pairs inside
                inner_text = value.strip('{} \n')
                sub_dict = {}
                # Split on commas outside of nested tables
                inner_parts = re.split(r',\s*(?![^{]*})', inner_text)
                for part in inner_parts:
                    if '=' in part:
                        inner_key, inner_value = part.split('=', 1)
                        inner_key = inner_key.strip()
                        inner_value = inner_value.strip()
                        sub_dict[inner_key] = convert_value(inner_value)
                result[key] = sub_dict
            else:
                result[key] = convert_value(value)
        
        return result
    
    return parse_table(lua_string)

def convert_to_yaml(lua_data):
    """Convert parsed Lua data to YAML structure."""
    yaml_data = {
        "character": {
            "gear": {f"slot{k}": v for k, v in lua_data["gear"].items()},
            "currencies": lua_data["currencies"]
        }
    }
    return yaml_data

def main():
    parser = argparse.ArgumentParser(description="Convert Lua exportData table to YAML")
    parser.add_argument("-f", required=True, help="Path to the Lua data file (e.g., export.lua)")
    parser.add_argument("-o", required=True, help="Path to the output YAML file (e.g., mychar.yaml)")
    args = parser.parse_args()

    # Read the Lua file
    try:
        with open(args.f, "r") as f:
            lua_content = f.read()
    except FileNotFoundError:
        print(f"Error: Input file '{args.f}' not found.")
        return
    except Exception as e:
        print(f"Error reading input file: {e}")
        return

    # Parse Lua to Python dictionary
    try:
        lua_data = parse_lua_table(lua_content)
        if "gear" not in lua_data or "currencies" not in lua_data:
            raise ValueError("Missing 'gear' or 'currencies' key in parsed Lua data")
    except Exception as e:
        print(f"Error parsing Lua data: {e}")
        return

    # Convert to YAML structure
    try:
        yaml_data = convert_to_yaml(lua_data)
    except Exception as e:
        print(f"Error converting to YAML: {e}")
        return

    # Write to YAML file
    try:
        with open(args.o, "w") as f:
            yaml.dump(yaml_data, f, default_flow_style=False, allow_unicode=True)
        print(f"YAML file '{args.o}' created successfully!")
    except Exception as e:
        print(f"Error writing YAML file: {e}")

if __name__ == "__main__":
    main()

File: convert-lua-string-to-yaml.rb
#!/usr/bin/env ruby

require 'yaml'
require 'optparse'

# Parse command-line arguments
options = {}
OptionParser.new do |opts|
  opts.banner = "Usage: ruby lua_to_yaml.rb -f INPUT_FILE -o OUTPUT_FILE"
  opts.on("-f FILE", "Path to the Lua data file (e.g., export.lua)") { |f| options[:input_file] = f }
  opts.on("-o FILE", "Path to the output YAML file (e.g., mychar.yaml)") { |o| options[:output_file] = o }
end.parse!

unless options[:input_file] && options[:output_file]
  puts "Error: Both -f and -o are required."
  exit 1
end

# Read the Lua file
begin
  lua_content = File.read(options[:input_file])
rescue Errno::ENOENT
  puts "Error: Input file '#{options[:input_file]}' not found."
  exit 1
rescue => e
  puts "Error reading input file: #{e}"
  exit 1
end

# Parse Lua table into Ruby hash
def parse_lua_table(text)
  text = text.sub(/exportData\s*=\s*/, '').strip
  
  def convert_value(val)
    val = val.strip
    return val.to_i if val =~ /^\d+$/
    return val if val == "N/A"
    val.gsub(/^"/, '').gsub(/"$/, '')
  end
  
  result = {}
  # Match top-level key-value pairs
  text.scan(/(\["[^"]+"\]|\d+)\s*=\s*({[^}]*}|[^,]+)(?:,|$|\s*\n*})/).each do |key, value|
    key = key.start_with?('[') ? key[2..-3] : key.to_i
    value = value.strip
    
    if value.start_with?('{')
      # Nested table
      sub_dict = {}
      value[1..-2].strip.scan(/([^=,\s]+)\s*=\s*("[^"]*"|[^,]+)(?:,|$)/).each do |inner_key, inner_value|
        sub_dict[inner_key] = convert_value(inner_value)
      end
      result[key] = sub_dict
    else
      result[key] = convert_value(value)
    end
  end
  
  result
end

# Convert to YAML structure
def convert_to_yaml(lua_data)
  {
    "character" => {
      "gear" => lua_data["gear"].transform_keys { |k| "slot#{k}" },
      "currencies" => lua_data["currencies"]
    }
  }
end

# Parse and convert
begin
  lua_data = parse_lua_table(lua_content)
  unless lua_data["gear"] && lua_data["currencies"]
    raise "Missing 'gear' or 'currencies' key in parsed Lua data"
  end
  yaml_data = convert_to_yaml(lua_data)
rescue => e
  puts "Error processing Lua data: #{e}"
  exit 1
end

# Write to YAML file
begin
  File.write(options[:output_file], yaml_data.to_yaml)
  puts "YAML file '#{options[:output_file]}' created successfully!"
rescue => e
  puts "Error writing YAML file: #{e}"
  exit 1
end